lib/connextclient/ConnextClient.ts

```diff
@@ -0,0 +1,699 @@
+import assert from 'assert';
+import http from 'http';
+import { SwapClientType, SwapRole, SwapState } from '../constants/enums';
+import { CurrencyInstance } from '../db/types';
+import Logger from '../Logger';
+import swapErrors from '../swaps/errors';
+import SwapClient, {
+  ChannelBalance,
+  ClientStatus,
+  PaymentState,
+  WalletBalance,
+  TradingLimits,
+  SwapClientInfo,
+} from '../swaps/SwapClient';
+import { SwapDeal } from '../swaps/types';
+import { UnitConverter } from '../utils/UnitConverter';
+import errors, { errorCodes } from './errors';
+import {
+  ConnextErrorResponse,
+  ConnextInitWalletResponse,
+  ConnextChannelConfig,
+  ConnextBalanceResponse,
+  ConnextTransferResponse,
+  ConnextChannelCount,
+  ConnextClientConfig,
+  ConnextInfo,
+  ConnextVersion,
+  TokenPaymentRequest,
+  ConnextTransferStatus,
+  ExpectedIncomingTransfer,
+  ProvidePreimageEvent,
+  TransferReceivedEvent,
+} from './types';
+import { parseResponseBody } from '../utils/utils';
+import { Observable, fromEvent } from 'rxjs';
+import { take, pluck, timeout, filter } from 'rxjs/operators';
+
+export const TIMELOCK_BUFFER = 100;
+
+const MAX_AMOUNT = Number.MAX_SAFE_INTEGER;
+
+interface ConnextClient {
+  on(event: 'preimage', listener: (preimageRequest: ProvidePreimageEvent) => void): void;
+  on(event: 'transferReceived', listener: (transferReceivedRequest: TransferReceivedEvent) => void): void;
+  on(event: 'htlcAccepted', listener: (rHash: string, amount: number, currency: string) => void): this;
+  on(event: 'connectionVerified', listener: (swapClientInfo: SwapClientInfo) => void): this;
+  once(event: 'initialized', listener: () => void): this;
+  emit(event: 'htlcAccepted', rHash: string, amount: number, currency: string): boolean;
+  emit(event: 'connectionVerified', swapClientInfo: SwapClientInfo): boolean;
+  emit(event: 'initialized'): boolean;
+  emit(event: 'preimage', preimageRequest: ProvidePreimageEvent): void;
+  emit(event: 'transferReceived', transferReceivedRequest: TransferReceivedEvent): void;
+}
+
+/**
+ * Waits for the preimage event from SwapClient for a specified hash
+ * @param client the swap client instance to listen events from
+ * @param expectedHash the expected hash of the payment
+ * @param errorTimeout optional maximum duration of time to wait for the preimage
+ */
+const waitForPreimageByHash = (
+  client: SwapClient,
+  expectedHash: string,
+  errorTimeout = 30000,
+): Promise<string> => {
+  // create an observable that emits values when a preimage
+  // event is triggered on the client
+  const preimage$: Observable<ProvidePreimageEvent> = fromEvent(client, 'preimage');
```

Comment:
We could do that, but then we'd have to to create the `preimage$` when starting/initializing the `ConnextClient` and attach it to the class itself:

```
private preimage$: Observable<ProvidePreimageEvent>;
...
this.preimage$ = fromEvent(this, 'preimage');
```
 
We could do that, but what would be the benefit of doing so? Keep in mind that observables by default are cold - meaning that it will not attach event listeners until there's a consumer:

```
observable.subscribe(...)
```

or 
```
observable.toPromise()
```